### DISCUSSION TITLE
Make LocalDirectory configurable

### DISCUSSION THREAD

**Discussion by williwyg** [dem-net/DEM.Net#93] [Ideas]
Bonjour Xavier,

Thanks for creating such a great library.

I trying to use your library for an application that is both hosted on Azure and a dedicated Windows server. 

I have some trouble however to get the ElevationService running on Azure because you have some hard-coded logic in the RasterService contructor to resolve the "LocalDirectory".  As soon as the RasterService class is instantiated on Azure the application crashes because the current implementation of the constructor is not valid on Azure. The path that is resolved is not writable since i use the deployment method "RunFromPackage". Which makes the path you are resolving readonly.

Also calling the "SetLocalDirectory" method is not an option because the RasterService class can not be instantiated.

A possible solution could be to make the "LocalDirectory" configurable through "DEMNetOptions". 

- Simply add a string property "LocalDirectory" to the DEMNetOptions class.
- Add the DEMNetOptions parameter to the constructor of RasterService class.
- Add a test to the RasterService constructor if the DEMNetOptions.LocalDirectory == null otherwise leave the logic as it is.

       public RasterService(
            RasterIndexServiceResolver rasterResolver, 
            IOptions<DEMNetOptions> options = null,
            ILogger<RasterService> logger = null
         )
        {
            this._logger = logger;
            this._rasterIndexServiceResolver = rasterResolver;

            _localDirectory = options?.Value.LocalDirectory ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), APP_NAME);

            if (!Directory.Exists(_localDirectory))
                Directory.CreateDirectory(_localDirectory);

            _metadataCatalogCache = new ConcurrentDictionary<string, List<FileMetadata>>();
        }

I am willing to donate if we can create a working solution.

Greetings,

Willy

**Comment by xfischer** [comment]
Hi @williwyg and thanks for pointing this out.

I have made a push on master branch. Could you test that out ?

Be sure to configure Options as : 
```csharp
var services = new ServiceCollection();
services.AddDemNetCore();
services.Configure<DEMNetOptions>(builder.GetSection(nameof(DEMNetOptions))); // <-- builder is IConfigurationRoot
// ...
```

**Comment by williwyg** [comment]
Hi Xavier,

Thanks for the quick response and fix.

I have tested the change you’ve made and after fiddling around a bit with the path setting I found the correct writable Azure web app path. “C:\local\temp\xxxxx”
Now I can nicely set the “LocalDirectory” setting in my Azure release pipeline for the Azure environment and leave the default behavior for my other environment.

When do you think this would be available in an official release.

And please tell me how much you like to get donated?

Greetings,

Willy



Van: Xavier Fischer ***@***.***>
Verzonden: maandag 3 mei 2021 12:51
Aan: dem-net/DEM.Net ***@***.***>
CC: Vroenhoven [Opinio] ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [dem-net/DEM.Net] Make LocalDirectory configurable (#93)


Hi @williwyg<https://github.com/williwyg> and thanks for pointing this out.

I have made a push on dev branch. Could you test that out ?

Be sure to configure Options as :

var services = new ServiceCollection();

services.AddDemNetCore();

services.Configure<DEMNetOptions>(builder.GetSection(nameof(DEMNetOptions))); // <-- builder is IConfigurationRoot

// ...

**Reply by xfischer** [reply]
Hi @williwyg, I hope this will work on Azure.
I'll release a NuGet in the next hours and keep you posted.
Regarding the donation, it's really up to you and the mirror you'll be facing. If you run a business out of this, then take the time you would spend coding the bits you use + the money your company will make and count 5% of the costs, that will give a reasonable estimate of the ceiling.
If you are doing this for personal use then it's really up to you and how the API sets inside your project (core or aside). That would range from 1$ to whatever amount you judge reasonable.

**Reply by xfischer** [reply]
@williwyg package 0.3.1 deployed :)

**Comment by williwyg** [comment]
Hi Xavier,

I have referenced the version 0.3.1 in my application and now it is working on Azure.

But unfortunately now it is broken in my other environment. This is due to a bug in the .net core Configuration handling as described here. https://github.com/dotnet/runtime/issues/36510
String properties are always deserialized as empty string instead of NULL.

  "DEMNetOptions": {
    "LocalDirectory": null
  }

Using the above config section leads to the following error. “Path cannot be the empty string or all whitespace” because in the RasterService constructor we check for null but the “LocalDirectory” property is deserialized with an empty string.

So I think the easiest way to handle this is to add an additional check for “!string.IsNullOrEmpty(options.Value.LocalDirectory)”.


Sorry for the inconvenience.

Greetings,

Willy






Van: Xavier Fischer ***@***.***>
Verzonden: maandag 3 mei 2021 17:31
Aan: dem-net/DEM.Net ***@***.***>
CC: Vroenhoven [Opinio] ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [dem-net/DEM.Net] Make LocalDirectory configurable (#93)


@williwyg<https://github.com/williwyg> package 0.3.1 deployed :)

**Reply by xfischer** [reply]
Fixed in 0.3.2, released on NuGet

**Comment by williwyg** [comment]
Wow. Didn’t expect this to be solved that fast.

Thanks.

Willy

Van: Xavier Fischer ***@***.***>
Verzonden: maandag 3 mei 2021 20:07
Aan: dem-net/DEM.Net ***@***.***>
CC: Vroenhoven [Opinio] ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [dem-net/DEM.Net] Make LocalDirectory configurable (#93)


Fixed in 0.3.2, released on NuGet

**Reply by xfischer** [reply]
Hi @williwyg, thanks for sponsoring me ! Let me know if you need any support, and if you'd like features to be added.

**Comment by williwyg** [comment]
Bonjour Xavier,

I have just started building an application to create routes for cyclists.

You might say routing applications are not a new thing and yes there are already a few available.

The problem with all available applications is they miss detailed information about the individual tracks.

So if I create a route with Garmin, Strava, etc.. I have no assurance that the route created is ridable and a good experience. There also is no possibility to differentiate between road and mtb.

They make use of there heatmaps where the fact that a lot of riders have driven a specific track is an indication it is a GOOD track??

I am creating an application that is based on OpenStreetMap data and where I am going to ask all cyclist to help me enhance the underlying data with more detailed information.

The principle will be like OpenStreetMap where everyone is encouraged to help enhance the data so eventually the generated routes will become better and better.

For a start it will be an open application like OpenStreetMap so I will not make any money with it.

I plan to work towards my retirement and this will keep me busy in my free time and coming winter evenings.


If I get stuck in the future and I believe you are the one to help me out I will turn to you.

Greetings,

Willy






Van: Xavier Fischer ***@***.***>
Verzonden: dinsdag 4 mei 2021 13:53
Aan: dem-net/DEM.Net ***@***.***>
CC: Vroenhoven [Opinio] ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [dem-net/DEM.Net] Make LocalDirectory configurable (#93)


Hi @williwyg<https://github.com/williwyg>, thanks for sponsoring me ! Let me know if you need any support, and if you'd like features to be added.

**Reply by xfischer** [reply]
Hi @williwyg and thanks for sharing this. This is a great idea, I hope you'll get it going and wish you success.
How do you plan to use elevations ? If it's for bike tracks, you may find `GetLineElevation` helpful, as it can simplify the resulting line.

**Comment by williwyg** [comment]
Hi Xavier,

Thanks for the hint. I will have a look on that method.

At the moment I am still experimenting and analyzing how to structure my data.

I divided Europa in small bounding boxes which I can activate individually.

Activating a boundingbox means importing the nodes and ways from OpenStreetMap and adding additional information.

Further I split up the Osm ways into segments. A segment is a sequence of osm nodes from one crossing to the next.

A Segment is my primary entity type to use for routing. The Segment will be enhanced with additional metadata like


  *   Length
  *   MaxGrade
  *   SurfaceType
  *   TotalMetersGain
  *   TotalMetersDrop
  *   SuitableForRoadRiding
  *   SuitableForMtbRiding
  *   Etc….

Some of these values can either be calculated or retrieved from existing resources. (Like elevation)
Other values will be provided by the users of the application

Eventually I want to have an api that can calculate a route based on parameters like

Type: MTB
Start: Lat,Lon
End: Lat, Lon
MaxGrade: 15%
Length: 60Km
MetersGain: between 1500 and 2000
MaxPercentageTarmac: 5%

I will start activating bounding boxes in the area where I live 50.28477104507804, 5.964175364966936

Greetings,

Willy



Van: Xavier Fischer ***@***.***>
Verzonden: woensdag 5 mei 2021 09:47
Aan: dem-net/DEM.Net ***@***.***>
CC: Vroenhoven [Opinio] ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [dem-net/DEM.Net] Make LocalDirectory configurable (#93)


Hi @williwyg<https://github.com/williwyg> and thanks for sharing this. This is a great idea, I hope you'll get it going and wish you success.
How do you plan to use elevations ? If it's for bike tracks, you may find GetLineElevation helpful, as it can simplify the resulting line.

**Reply by xfischer** [reply]
Did a lot of experiments with OSM data also (I have generated vector tiles with elevations for every OSM node, might be useful).
How do you retrieve data from OSM ? Overpass or PBF ?

**Comment by williwyg** [comment]
Hi Xavier,

I am using Overpass api to retrieve Osm nodes and ways in json format.

I am relatively new in working with OSM and map software so I am not sure if I am doing the things right.

As I already said I still experimenting and learning.

It would be nice if you could advise me when I get stuck.

Greetings,

Willy

Van: Xavier Fischer ***@***.***>
Verzonden: woensdag 5 mei 2021 10:38
Aan: dem-net/DEM.Net ***@***.***>
CC: Vroenhoven [Opinio] ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [dem-net/DEM.Net] Make LocalDirectory configurable (#93)


Did a lot of experiments with OSM data also (I have generated vector tiles with elevations for every OSM node, might be useful).
How do you retrieve data from OSM ? Overpass or PBF ?

**Comment by xfischer** [comment]
Yes, sure !
OverpassAPI is great but is rate limited, so it cannot be used in a production app where you expect proxying many requests from users. It is still OK if you want to preprocess the data.
Then regarding OSM itself you may find bike-related info on Osm wiki, I think there is a lot work done.

PBF files are raw data, expensive to compute but very useful : you have all the information.



________________________________
De : williwyg ***@***.***>
Envoyé : mercredi 5 mai 2021 10:49
À : dem-net/DEM.Net ***@***.***>
Cc : Xavier Fischer ***@***.***>; Comment ***@***.***>
Objet : Re: [dem-net/DEM.Net] Make LocalDirectory configurable (#93)


Hi Xavier,

I am using Overpass api to retrieve Osm nodes and ways in json format.

I am relatively new in working with OSM and map software so I am not sure if I am doing the things right.

As I already said I still experimenting and learning.

It would be nice if you could advise me when I get stuck.

Greetings,

Willy

Van: Xavier Fischer ***@***.***>
Verzonden: woensdag 5 mei 2021 10:38
Aan: dem-net/DEM.Net ***@***.***>
CC: Vroenhoven [Opinio] ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [dem-net/DEM.Net] Make LocalDirectory configurable (#93)


Did a lot of experiments with OSM data also (I have generated vector tiles with elevations for every OSM node, might be useful).
How do you retrieve data from OSM ? Overpass or PBF ?

**Comment by williwyg** [comment]
Hi Xavier,

I do not know if you could give me some advice on map software.

Currently for presenting maps in my application I make use of Leaflet. So far I am happy with the features Leaflet provides. But I am not sure for the longer term.

Do you have any experience using Leaflet or Mapbox. If so which one would you advise. What are the criteria to choose one over the other.

Greetings,

Willy

**Reply by xfischer** [reply]
Hi @williwyg, I'd say Leaflet will get you covered with the basics, but MapBox has way more functionality.

It's not licensed the same, so you'd better check the "libre" alternative here : https://github.com/maplibre/maplibre-gl-js

There are nice examples here : https://maplibre.org/maplibre-gl-js-docs/example/
